### PR TITLE
Support unregister device

### PR DESCRIPTION
**Pull request by ben181231**
connect SkygearIO/skygear-server#245
connect SkygearIO/skygear-server#249

### CODE REVIEW

**Comment by oursky-travis** [comment]
@ben181231, thanks for your PR! By analyzing the history of the files in this pull request, we identified @cheungpat and @rickmak to be potential reviewers.

File: Pod/Classes/SKYContainer.m [hunk 1]
```diff
@@ -642,6 +653,29 @@ - (void)registerDeviceCompletionHandler:(void (^)(NSString *, NSError *))complet
...
+}
+
+- (void)unregisterDeviceCompletionHandler:(void (^)(NSString *deviceID,
+                                                    NSError *error))completionHandler
+{
+    NSString *existingDeviceID = self.registeredDeviceID;
+    if (existingDeviceID != nil) {
+        SKYUnregisterDeviceOperation *operation =
+            [SKYUnregisterDeviceOperation operationWithDeviceID:existingDeviceID];
+        operation.unregisterCompletionBlock = completionHandler;
```
Comment (cheungpat) [inline Pod/Classes/SKYContainer.m]: The `completionHandler` has to be called in the main thread. (do not fix it in the Operation)

File: Pod/Classes/Operations/SKYUnregisterDeviceOperation.m [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+{
+    self = [super init];
+    if (self) {
+        self.deviceID = [deviceID copy];
+    }
+
+    return self;
+}
+
+- (void)setUnregisterCompletionBlock:(void (^)(NSString *, NSError *))unregisterCompletionBlock
```
Comment (cheungpat) [inline Pod/Classes/Operations/SKYUnregisterDeviceOperation.m]: While this certainly works, I think the better style to follow is in SKYFetchRecordsOperation, in which we do not override the setter for the completion block.

#justsaying

File: Pod/Classes/SKYContainer.h [hunk 3]
```diff
@@ -214,6 +214,20 @@ typedef void (^SKYContainerUserOperationActionCompletion)(SKYUser *user, NSError
...
+ */
+- (void)unregisterDevice;
+
+/**
+ * Unregister the current user from the current device.
+ *
+ * @param completionHandler the completion handler
+ *
+ */
+- (void)unregisterDeviceCompletionHandler:(void (^)(NSString *deviceID,
```
Comment (cheungpat) [inline Pod/Classes/SKYContainer.h]: you need a test case for this method in `SKYContainerTests.m`

File: Pod/Classes/SKYContainer.m [hunk 4]
```diff
@@ -642,6 +653,29 @@ - (void)registerDeviceCompletionHandler:(void (^)(NSString *, NSError *))complet
                               }];
 }
 
+- (void)unregisterDevice
+{
+    [self unregisterDeviceCompletionHandler:^(NSString *deviceID, NSError *error) {
+        if (error != nil) {
+            NSLog(@"Warning: Fail to unregister device: %@", error.localizedDescription);
```
Comment (cheungpat) [inline Pod/Classes/SKYContainer.m]: s/Fail/Failed/

File: Pod/Classes/SKYContainer.h (hunk 3 above)
Comment (ben181231) [inline Pod/Classes/SKYContainer.h]: Test cases added.

File: Pod/Classes/SKYContainer.m (hunk 1 above)
Comment (ben181231) [inline Pod/Classes/SKYContainer.m]: Changed to dispatch to main queue.

**Comment by ben181231** [comment]
The PR is updated.

**Pull request merged**